_includes/markdown/Typescript.md

```diff
@@ -0,0 +1,188 @@
+
+## When to use TypeScript
+
+At 10up we require that all projects considered "JavaScript-First" to make use of TypeScript. Generally speaking Standard WordPress projects (including block development) would not fall under "JavaScript-First" projects. 
+
+If you are building a project using Next.js and/or 10up's [HeadstartWP framework](https://headstartwp.10up.com/), Sanity and any other headless CMS'es powered by a React front-end then we require TypeScript unless other circumstances make using just JavaScript a better choice.
+
+The main reason we don't require TypeScript for standard WordPress projects at the moment is that TypeScript support for block development is still very limited, largely driven by community types packages that don't always match the actual types for the `@wordpress` packages.
+
+## How to use TypeScript
+
+In an ideal world, every engineer would be fully capable of writing all sorts of complex TypeScript types, however, we understand that it takes time to master a language like TypeScript, especially in a few scenarios that require writing more complex types. Therefore, while we use a reasonably strict `tsconfig.json` we are flexible in practice.
+
+We want the TypeScript to always be valid and issue no errors at compilation, but we allow a few escape hatches should engineers get stuck with a particular typing problem. We follow the principles below when writing TypeScript:
+
+### 1. Never ship code that has compilation issues
+
+Depending on the tooling you're using (e.g. Next.js) it is possible to disable type checking and still compile invalid TS code as long as the underlying JS code is valid. While the runtime behavior might still be working as expected this is not a recommended practice and generally would just make engineers not have trust in the type system. If you are facing a challenge typing something it is better to use some of the escape hatches documented below instead of just ignoring TS errors.
+
+### 2. Favor efficiency instead of strict typing
+In an ideal world, all engineers would be capable of typing everything properly from the get-go. However, we know that is not the case and engineers need time and practice to get there. For that reason, we favor efficiency over strict typing. 
+
+Our philosophy is that over time engineers will organically get more familiar with TypeScript and strictness will come as a result. Code reviews are also a place where engineers might get quick feedback that would allow them to more efficiently address typing issues instead of wasting a lot of time during initial development trying to figure it out on their own.
+
+When facing such typing challenges we recommend the following strategies as an escape hatch:
+
+#### 1. The `as unknown as ExpectedType` type cast
+
+When struggling with a type that does not match the expected type, one of the escape hatches we recommend is to forcibly typecast the type to the expected type:
+
+```ts
```

Comment:
I ❤️ TODOs. I'm a big fan of engineers using them and also leaving their name with it so they take ownership of it `TODO(tobeycodes): Something...` which I think is important so they come back and fix it later. So often TODOs are left in the codebase forever. 

This is a great plugin for VSCode users to help manage them too https://marketplace.visualstudio.com/items?itemName=Gruntfuggly.todo-tree

Reply:
what if we could use eslint to detect todos and show as warnings 🤔 that wound't prevent todos from being left in the codebase forever but would be an easy way to list them.

Reply:
Actually I think a better think if we encourage todos is to have a eslint rule to enforce the `TODO(name)` pattern. I don't think it makes sense for the todo itself to be a warning.

Reply:
I actually do not like warnings. I think they create a lot of noise and more confusion for engineers where they are not sure whether they need to make a change or if it is ok to leave their code as it is. I'd rather make it clear to engineers what needs to be fixed and otherwise leave their code. Similar to TODOs I don't think warnings ever get cleaned up.

https://www.npmjs.com/package/eslint-plugin-only-warn

Reply:
Added a note about the TODOs

Reply:
Great, thanks @nicholasio @tobeycodes - I like the idea of eslint handling this.